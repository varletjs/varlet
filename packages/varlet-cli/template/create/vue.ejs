<template>
  <div class="var-<%- kebabCaseName %>"></div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import './<%- kebabCaseName %>.less'

export default defineComponent({
  name: '<%- namespace + bigCamelizeName %>',
})
</script>
