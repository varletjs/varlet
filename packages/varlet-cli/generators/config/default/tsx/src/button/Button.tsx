import { defineComponent, PropType } from 'vue'
import './button.less'

export default defineComponent({
  name: 'BButton',
  props: {
    color: {
      type: String
    },
    onClick: {
      type: Function as PropType<(e: Event) => void>
    }
  },
  setup(props, { slots }) {
    const handleClick = (e: Event) => props.onClick?.(e)

    return () => {
      const { color } = props

      return <>
        <button
          class="basic-button"
          style={{ background: color }}
          onClick={ handleClick }
        >
          { slots.default?.() }
        </button>
      </>
    }
  }
})
