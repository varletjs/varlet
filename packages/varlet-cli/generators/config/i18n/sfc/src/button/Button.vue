<template>
  <button class="basic-button" :style="{ background: color }">
    <slot />
<!--    {{ pack.button }}-->
  </button>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue'

// i18n for component's internal
// import { pack } from '../locale'

export default defineComponent({
  name: 'BButton',
  props: {
    color: {
      type: String
    },
    onClick: {
      type: Function as PropType<(e: Event) => void>
    }
  },
  setup(props) {
    const handleClick = (e: Event) => props.onClick?.(e)

    return {
      handleClick
    }
  }
})
</script>

<style lang="less">
@import "button";
</style>
