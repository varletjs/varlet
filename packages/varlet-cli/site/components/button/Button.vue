<template>
  <button
    v-ripple="{ disabled: disabled || !ripple }"
    :class="
      classes(
        n(),
        'var-site--box',
        n(`--${size}`),
        [block, `var-site--flex ${n('--block')}`, 'var-site--inline-flex'],
        [disabled, n('--disabled')],
        [text, `${n(`--text-${type}`)} ${n('--text')}`, `${n(`--${type}`)} var-site-elevation--2`],
        [text && disabled, n('--text-disabled')],
        [round, n('--round')],
        [outline, n('--outline')]
      )
    "
    :style="{
      color: textColor,
      background: color,
    }"
    :disabled="disabled"
    @click="handleClick"
    @touchstart="handleTouchstart"
  >
    <var-loading
      :class="n('loading')"
      var-site-button-cover
      :color="loadingColor"
      :type="loadingType"
      :size="loadingSize"
      :radius="loadingRadius"
      v-if="loading || pending"
    />
    <div :class="classes(n('content'), [loading || pending, n('--hidden')])">
      <slot />
    </div>
  </button>
</template>

<script lang="ts">
import Ripple from '../ripple'
import VarLoading from '../loading'
import { defineComponent, ref } from 'vue'
import { props } from './props'
import { createNamespace } from '../utils/components'
import type { Ref } from 'vue'

const { n, classes } = createNamespace('button')

export default defineComponent({
  name: 'VarButton',
  components: {
    VarLoading,
  },
  directives: { Ripple },
  props,
  setup(props) {
    const pending: Ref<boolean> = ref(false)

    const attemptAutoLoading = (result: any) => {
      if (props.autoLoading) {
        pending.value = true
        Promise.resolve(result).finally(() => {
          pending.value = false
        })
      }
    }

    const handleClick = (e: Event) => {
      const { loading, disabled, onClick } = props

      if (!onClick || loading || disabled || pending.value) {
        return
      }

      attemptAutoLoading(onClick(e))
    }

    const handleTouchstart = (e: Event) => {
      const { loading, disabled, onTouchstart } = props

      if (!onTouchstart || loading || disabled || pending.value) {
        return
      }

      attemptAutoLoading(onTouchstart(e))
    }

    return {
      n,
      classes,
      pending,
      handleClick,
      handleTouchstart,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../styles/elevation';
@import '../ripple/ripple';
@import '../loading/loading';
@import './button';
</style>
