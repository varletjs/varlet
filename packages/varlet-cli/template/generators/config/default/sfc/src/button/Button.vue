<template>
  <button class="basic-button" :style="{ background: color }">
    <slot />
  </button>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue'
export default defineComponent({
  name: 'BButton',
  props: {
    color: {
      type: String,
    },
    onClick: {
      type: Function as PropType<(e: Event) => void>,
    },
  },
  setup(props) {
    const handleClick = (e: Event) => props.onClick?.(e)
    return {
      handleClick,
    }
  },
})
</script>

<style lang="less">
@import './button';
</style>
