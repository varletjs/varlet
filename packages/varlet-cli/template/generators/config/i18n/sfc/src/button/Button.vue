<template>
  <button class="var-button" :style="{ background: color }">
    {{ pack.button }}
    <slot />
  </button>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue'

// i18n for component's internal
import { pack } from '../locale'

export default defineComponent({
  name: 'VarButton',
  props: {
    color: {
      type: String,
    },
    onClick: {
      type: Function as PropType<(e: Event) => void>,
    },
  },
  setup(props) {
    const handleClick = (e: Event) => props.onClick?.(e)

    return {
      pack,
      handleClick,
    }
  },
})
</script>

<style lang="less">
@import './button';
</style>
