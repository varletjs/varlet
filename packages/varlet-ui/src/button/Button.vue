<template>
  <button
    v-ripple="{ disabled: disabled || !ripple }"
    class="var-button var--box"
    :class="[
      `var-button--${size}`,
      block ? 'var--flex var-button--block' : 'var--inline-flex',
      disabled ? 'var-button--disabled' : null,
      text ? `var-button--text-${type}` : `var-button--${type}`,
      text ? 'var-button--text' : 'var-elevation--2',
      text && disabled ? 'var-button--text-disabled' : null,
      round ? 'var-button--round' : null,
      outline ? 'var-button--outline' : null,
    ]"
    :style="{
      color: textColor,
      background: color,
    }"
    :disabled="disabled"
    @click="handleClick"
    @touchstart="handleTouchstart"
  >
    <var-loading
      class="var-button__loading"
      var-button-cover
      :type="loadingType"
      :size="loadingSize"
      :radius="loadingRadius"
      v-if="loading || pending"
    />
    <div class="var-button__content" :class="[loading || pending ? 'var-button--hidden' : null]">
      <slot />
    </div>
  </button>
</template>

<script lang="ts">
import Ripple from '../ripple'
import VarLoading from '../loading'
import { defineComponent, Ref, ref } from 'vue'
import { props } from './props'
import { isPromise } from '../utils/shared'

export default defineComponent({
  name: 'VarButton',
  components: {
    VarLoading,
  },
  directives: { Ripple },
  props,
  setup(props) {
    const pending: Ref<boolean> = ref(false)

    const attemptAutoLoading = (result: any) => {
      if (props.autoLoading) {
        pending.value = true
        Promise.resolve(result).finally(() => {
          pending.value = false
        })
      }
    }

    const handleClick = (e: Event) => {
      const { loading, disabled, onClick } = props

      if (!onClick || loading || disabled || pending.value) {
        return
      }

      attemptAutoLoading(onClick(e))
    }

    const handleTouchstart = (e: Event) => {
      const { loading, disabled, onTouchstart } = props

      if (!onTouchstart || loading || disabled || pending.value) {
        return
      }

      attemptAutoLoading(onTouchstart(e))
    }

    return {
      pending,
      handleClick,
      handleTouchstart,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../styles/elevation';
@import '../ripple/ripple';
@import '../loading/loading';
@import './button';
</style>
