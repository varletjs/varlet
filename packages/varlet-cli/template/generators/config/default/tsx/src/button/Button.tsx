import { defineComponent, PropType } from 'vue'
import './button.less'

export default defineComponent({
  name: 'VarButton',
  props: {
    color: {
      type: String,
    },
    onClick: {
      type: Function as PropType<(e: Event) => void>,
    },
  },
  setup(props, { slots }) {
    const handleClick = (e: Event) => props.onClick?.(e)

    return () => {
      const { color } = props

      return (
        <>
          <button class="var-button" style={{ background: color }} onClick={handleClick}>
            {slots.default?.()}
          </button>
        </>
      )
    }
  },
})
