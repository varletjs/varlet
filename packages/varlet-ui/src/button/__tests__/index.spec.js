import example from '../example'
import Button from '..'
import VarButton from '../Button'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'

test('test button example', () => {
  const wrapper = mount(example)
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.unmount()
})

test('test button plugin', () => {
  const app = createApp({}).use(Button)
  expect(app.component(Button.name)).toBeTruthy()
})

test('test button onClick & onTouchstart null callback', () => {
  const wrapper = mount(VarButton)
  wrapper.trigger('click')
  wrapper.trigger('touchstart')
  wrapper.unmount()
})

test('test button onClick & onTouchstart', () => {
  const onClick = jest.fn()
  const onTouchstart = jest.fn()

  const wrapper = mount(VarButton, {
    props: {
      onClick,
      onTouchstart,
    },
  })

  wrapper.trigger('click')
  wrapper.trigger('touchstart')
  expect(onClick).toHaveBeenCalledTimes(1)
  expect(onTouchstart).toHaveBeenCalledTimes(1)
  wrapper.unmount()
})

test('test button disabled', () => {
  const onClick = jest.fn()
  const onTouchstart = jest.fn()

  const wrapper = mount(VarButton, {
    props: {
      disabled: true,
      onClick,
      onTouchstart,
    },
  })
  expect(wrapper.props('disabled')).toBeTruthy()
  wrapper.trigger('click')
  wrapper.trigger('touchstart')
  expect(onClick).toHaveBeenCalledTimes(0)
  expect(onTouchstart).toHaveBeenCalledTimes(0)
  wrapper.unmount()
})

test('test button loading', () => {
  const onClick = jest.fn()
  const onTouchstart = jest.fn()

  const wrapper = mount(VarButton, {
    props: {
      loading: true,
    },
  })
  wrapper.trigger('click')
  wrapper.trigger('touchstart')
  expect(wrapper.props('loading')).toBeTruthy()
  expect(onClick).toHaveBeenCalledTimes(0)
  expect(onTouchstart).toHaveBeenCalledTimes(0)
  wrapper.unmount()
})

test('test button color & textColor', () => {
  const wrapper = mount(VarButton, {
    props: {
      color: '#000',
      textColor: '#000',
    },
  })
  expect(wrapper.props('color')).toBe('#000')
  expect(wrapper.props('textColor')).toBe('#000')
  wrapper.unmount()
})

test('test button styles', () => {
  const wrapper = mount(VarButton, {
    props: {
      size: 'large',
      round: true,
      text: true,
      block: true,
    },
    slots: {
      default: () => 'button text',
    },
    
  })
  expect(wrapper.classes()).toContain('var-button--round')
  expect(wrapper.classes()).toContain('var-button--text')
  expect(wrapper.classes()).toContain('var-button--block')
  expect(wrapper.classes()).toContain('var-button--large')
  expect(wrapper.text()).toBe('button text')
  wrapper.unmount()
})
