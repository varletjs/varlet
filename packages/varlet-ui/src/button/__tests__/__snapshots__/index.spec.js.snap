// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test button group color and text-color 1`] = `
"<div class=\\"var-button-group var--box var-button-group--mode-normal var-button-group--horizontal var-elevation--2\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--0\\" style=\\"color: red; background: yellow;\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--0\\" style=\\"color: red; background: yellow;\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--0\\" style=\\"color: red; background: yellow;\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button></div>"
`;

exports[`test button group default slots 1`] = `
"<div class=\\"var-button-group var--box var-button-group--mode-normal var-button-group--horizontal var-elevation--2\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--0\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--0\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--0\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button></div>"
`;

exports[`test button group elevation 1`] = `
"<div class=\\"var-button-group var--box var-button-group--mode-normal var-button-group--horizontal var-elevation--10\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--0\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--0\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--0\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button></div>"
`;

exports[`test button group outline mode 1`] = `
"<div class=\\"var-button-group var--box var-button-group--mode-outline var-button-group--horizontal\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-button--outline\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-button--outline\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text var-button--outline\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button></div>"
`;

exports[`test button group text mode 1`] = `
"<div class=\\"var-button-group var--box var-button-group--mode-text var-button-group--horizontal\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-default var-button--text\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button></div>"
`;

exports[`test button group type and size 1`] = `
"<div class=\\"var-button-group var--box var-button-group--mode-normal var-button-group--horizontal var-elevation--2\\"><button class=\\"var-button var--box var-button--large var--inline-flex var-button--primary var-elevation--0\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button><button class=\\"var-button var--box var-button--large var--inline-flex var-button--primary var-elevation--0\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button><button class=\\"var-button var--box var-button--large var--inline-flex var-button--primary var-elevation--0\\" type=\\"button\\">
    <!--v-if-->
    <div class=\\"var-button__content\\"></div>
    <div class=\\"var-button__overlay\\"></div>
  </button></div>"
`;
